homeassistant/util/timeout.py

```diff
@@ -0,0 +1,457 @@
+    def freezes_done(self) -> bool:
+        """Return True if all freeze are done."""
+        return len(self._freezes) == 0 and self._manager.freezes_done
+
+    def enter_task(self, task: _TaskZone) -> None:
+        """Start into new Task."""
+        if self.state == _State.TIMEOUT:
+            raise asyncio.TimeoutError
+        elif self.state != _State.ENTER:
+            self._start_timer()
+
+        self._count += 1
+        self._tasks.append(task)
+
+    def exit_task(self, task: _TaskZone, exc_type: Type[BaseException]) -> None:
+        """Exit a running Task."""
+        self._count -= 1
+
+        # Timeout exit
+        if exc_type is asyncio.CancelledError and self.state == _State.TIMEOUT:
+            if self._count == 0:
+                self._manager.drop_zone(self.name)
+            raise asyncio.TimeoutError
+
+        # On latest listener
+        if self._count == 0 and self.state != _State.EXIT:
+            self._state = _State.EXIT
+            self._stop_timer()
+            self._manager.drop_zone(self.name)
+
+        self._tasks.remove(task)
+
+    def _start_timer(self) -> None:
+        """Start timeout handler."""
+        self._state = _State.ENTER
+        self._time = self._loop.time() + self._timeout
+        self._timeout_handler = self._loop.call_at(self._time, self._on_timeout)
+
+    def _stop_timer(self) -> None:
+        """Stop zone timer."""
+        if self._timeout_handler is None:
+            return
+
+        self._timeout_handler.cancel()
+        self._timeout_handler = None
+        # Calculate new timeout
+        self._timeout = self._time - self._loop.time()
+
+    def _on_timeout(self) -> None:
+        """Process timeout."""
+        self._timeout_handler = None
+        self._state = _State.TIMEOUT
+
+        # Cancel all running tasks
+        for task in self._tasks:
+            if task.done():
+                continue
+            task.cancel()
+
+    def pause(self) -> None:
+        """Stop timers while it freeze."""
+        if self._state != _State.ENTER:
+            return
+        self._state = _State.FREEZE
+        self._stop_timer()
+
+    def reset(self) -> None:
+        """Reset timer after freeze."""
+        if self._state != _State.FREEZE:
+            return
+        self._state = _State.ENTER
+        self._start_timer()
+
+
+class ZoneTimeout:
+    """Async zone based timeout handler."""
+
+    def __init__(self):
+        """Initalize ZoneTimeout handler."""
+        self._loop: asyncio.AbstractEventLoop = asyncio.get_running_loop()
+        self._zones: Dict[str, _Zone] = {}
+        self._globals: List[_TaskGlobal] = []
+        self._freezes: List[_FreezeGlobal] = []
+
+    @property
+    def zones_done(self) -> bool:
+        """Return True if all zones are finish."""
+        return not bool(self._zones)
+
+    @property
+    def freezes_done(self) -> bool:
+        """Return True if all freezes are finish."""
+        return not bool(self._freezes)
```

Comment:
No need to wrap in `bool(…)` as `not` is a boolean operator which will already call `self._freezes.__bool__()`
```suggestion
        return not self._freezes
```